File: TypeScriptAndAngularProjects/testing/exercises/src/gameOfLife/cell.js
export class GameOfLifeCell {

    constructor(neighbours) {
        if (!(this instanceof GameOfLifeCell)) {
            return new GameOfLifeCell(neighbours);
        }
        const numNeighbours = neighbours.length;
        if (numNeighbours > 8) {
            throw "Too many neighbours";
        }
        this._neighbours = neighbours;
        this._state = "dead";
    }

    toString() {
        return "A Game Of Life Cell";
    }

    getState() {
        return this._state;
    }

    makeDead() {
        this._state = "dead";
    }

    makeAlive() {
        this._state = "alive";
    }

    changeState() {
        function countAliveNeighbours(neighbours) {
            var count = 0;
            neighbours.forEach(function (item) {
                if (item.getState() === "alive") {
                    count++;
                }
            });
            return count;
        }

        var count = countAliveNeighbours(this._neighbours);
        if (this._state === "dead") {
            if (count === 3) {
                this.makeAlive();
            }
        } else if (this._state === "alive") {
            if (count > 3 || count < 2) {
                this.makeDead();
            }
        }
    }
}

File: JavaProjects/JavaLanguageExercises/src/game/of/life/finish/Cell.java
package game.of.life.finish;

import java.util.List;
import java.util.stream.Collectors;

import static game.of.life.finish.CellState.ALIVE;
import static game.of.life.finish.CellState.DEAD;

public class Cell {
    private List<Cell> neighbours;
    private CellState state;

    public Cell(List<Cell> neighbours) {
        super();
        state = DEAD;
        this.neighbours = neighbours;
        if (neighbours.size() > 8) {
            throw new IllegalStateException();
        }
    }

    public void makeAlive() {
        setState(ALIVE);
    }

    public void makeDead() {
        setState(DEAD);
    }

    private void setState(CellState newState) {
        state = newState;
    }

    public void addNeighbour(Cell neighbour) {
        if (neighbours.size() == 8) {
            throw new IllegalStateException();
        }
        neighbours.add(neighbour);
    }

    public void changeState() {
        long aliveNeighbours = countAliveNeighbours();
        CellState result = DEAD;
        if (state == DEAD && aliveNeighbours == 3) {
            result = ALIVE;
        } else if (state == ALIVE) {
            if (aliveNeighbours == 2 || aliveNeighbours == 3) {
                result = ALIVE;
            }
        }
        state = result;
    }

    private long countAliveNeighbours() {
        return neighbours.stream()
                .filter(Cell::isAlive)
                .collect(Collectors.counting());
    }

    public boolean isAlive() {
        return state == ALIVE;
    }

    public boolean isDead() {
        return state == DEAD;
    }
}